### PR TITLE
dependency checking

### PR DESCRIPTION
**Pull request by jxa**
- check for required inputs
- support optional arguments (with default values)

### CODE REVIEW

**Comment by jxa** [comment]
RFR @tonywok

File: lib/funk/input_fn.rb [hunk 1]
```diff
@@ -1,15 +1,34 @@
...
     def call(input)
-      input[name]
+      if input.has_key?(name)
+        input[name]
+      else
+        NO_INPUT_PROVIDED
+      end
+    end
+
+    def required?
```
Comment (tonywok) [inline lib/funk/input_fn.rb]: TODO: remove first attempt code.

File: lib/funk/input_fn.rb [hunk 2]
```diff
@@ -1,15 +1,34 @@
...
+      @required
+    end
+
+    def require!
```
Comment (tonywok) [inline lib/funk/input_fn.rb]: TODO: remove first attempt code.

File: lib/funk/input_fn.rb [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
+  def NO_INPUT_PROVIDED.inspect
+    "[not provided]"
+  end
+
   class InputFn < Fn
 
     def initialize(name, _=nil)
       @name = name
       @dependencies = []
+      @required = false
```
Comment (tonywok) [inline lib/funk/input_fn.rb]: TODO: remove first attempt code.

File: lib/funk/input_fn.rb [hunk 4]
```diff
@@ -1,15 +1,34 @@
 require "funk/fn"
+require "funk/missing_dependencies_exception"
 
 module Funk
+  NO_INPUT_PROVIDED = Object.new
```
Comment (tonywok) [inline lib/funk/input_fn.rb:5]: document optional argument in diff function edgecase

**Pull request merged**